Batcher implementation with Synchronized `Batcher#flush`(without triggers)

rahulKQL opened the pull request:
This is a follow-up PR for #692 

## What this PR contains
- This implementation **does not include any triggers/thresholds/flowController** for sending received elements for batching(This will be added in followup PRs).
- As of this implementation, The User has to explicitly call the `Batcher#flush()` to send the accumulated request for batching.
- Here `BatcherImpl` expects `v2.BatchingDescriptor`, `UnaryCallable` and a `request prototype` containing repetitive data which would be copied over to each batch along with the bundle of elements.
- `v2.BatchingCallSettings` is an extension from existing `BatchingCallSettings`.
- Have added default wait time for unfinished batch as `250ms`. 


### Follow-ups
Once this change is merged I will raise a separate PR with triggers, thresholds, and flowController for `BatcherImpl`.

rahulKQL commented:
@igorbernstein2 Please have a look.

igorbernstein2 commented on gax/src/main/java/com/google/api/gax/batching/v2/BatcherImpl.java:
```diff
@@ -0,0 +1,189 @@
+import java.util.concurrent.locks.LockSupport;
+
+/**
+ * Queues up the elements until {@link #flush()} is called, Once batching is finished returned
```
s/Once/once/

igorbernstein2 commented on gax/src/main/java/com/google/api/gax/batching/v2/BatcherImpl.java:
```diff
@@ -0,0 +1,189 @@
+ *
+ * <p>This class is not thread-safe, and requires calling classes to make it thread safe.
+ */
+public class BatcherImpl<ElementT, ElementResultT, RequestT, ResponseT>
```
`@BetaApi("Batching surface is not yet stable")`

igorbernstein2 commented on gax/src/main/java/com/google/api/gax/batching/v2/BatcherImpl.java:
```diff
@@ -0,0 +1,189 @@
+ * Queues up the elements until {@link #flush()} is called, Once batching is finished returned
+ * future gets resolves.
+ *
+ * <p>This class is not thread-safe, and requires calling classes to make it thread safe.
```
``` <p>This class is not thread-safe, and expects to be used from a single thread```

igorbernstein2 commented on gax/src/main/java/com/google/api/gax/batching/v2/BatcherImpl.java:
```diff
@@ -0,0 +1,189 @@
+    sendBatch();
+    while (numOfRpcs.get() > 0) {
+      LockSupport.parkNanos(DEFAULT_FINISH_WAIT_NANOS);
+    }
```
I don't love this. Can we use a semaphore instead?

rahulKQL replied:
I have added a semaphore implementation 0 permits. Please let me know if this impl looks good  to you 
`private final AtomicInteger numOfRpcs = new AtomicInteger(0);`

rahulKQL replied:
Thanks for the detailed steps, I have implemented with the same for `Batcher#flush()`. Please have a look.

igorbernstein2 commented on gax/src/main/java/com/google/api/gax/batching/v2/BatcherImpl.java:
```diff
@@ -0,0 +1,189 @@
+    }
+  }
+
+  /** sends accumulated elements asynchronously for batching. */
```
s/sends/Sends

igorbernstein2 commented on gax/src/main/java/com/google/api/gax/batching/v2/BatcherImpl.java:
```diff
@@ -0,0 +1,189 @@
+  /** {@inheritDoc} */
+  @Override
+  public ApiFuture<ElementResultT> add(final ElementT element) {
+    Preconditions.checkState(!isClosed, "Cannot perform batching on a closed connection");
```
This has nothing to do with a connection.

`Can't add elements to a closed a batcher`

igorbernstein2 commented on gax/src/main/java/com/google/api/gax/batching/v2/BatcherImpl.java:
```diff
@@ -0,0 +1,189 @@
+  private boolean isClosed = false;
+
+  private BatcherImpl(Builder<ElementT, ElementResultT, RequestT, ResponseT> builder) {
+    this.prototype = Preconditions.checkNotNull(builder.prototype);
```
Please add messages indicating which attribute is missing so that the caller doesn't have to dig through the code here to figure out what they forgot to set

igorbernstein2 commented on gax/src/main/java/com/google/api/gax/batching/v2/BatchingCallSettings.java:
```diff
@@ -0,0 +1,141 @@
+import java.util.Set;
+
+/**
+ * This is an extension to {@link com.google.api.gax.rpc.BatchingCallSettings}, to support for
```
Please don't refer to v1 classes

igorbernstein2 commented on gax/src/main/java/com/google/api/gax/batching/v2/BatcherImpl.java:
```diff
@@ -0,0 +1,189 @@
+   * This class represent one logical Batch. It accumulates all the elements and it's corresponding
+   * future element results for one batch.
+   */
+  class Batch {
```
static?

rahulKQL replied:
My intention was not to declare type params for inner class... but Have updated it to `private static class`.

igorbernstein2 replied:
Please add a comment to make that clear

igorbernstein2 commented on gax/src/test/java/com/google/api/gax/batching/v2/BatcherImplTest.java:
```diff
@@ -0,0 +1,197 @@
+    } catch (InterruptedException | ExecutionException e) {
+      Truth.assertThat(e.getCause()).isInstanceOf(RuntimeException.class);
+      throw e;
+    }
```
Please use something like:
https://github.com/googleapis/google-cloud-java/blob/master/google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/admin/v2/stub/AwaitReplicationCallableTest.java#L97-L105To test exceptions, it makes it easier to see whats happening as opposed using a combination of expected annotations and rethrowing it.

Also, please unwrap the ExecutionException and verify that the cause is expected

igorbernstein2 commented on gax/src/test/java/com/google/api/gax/batching/v2/BatcherImplTest.java:
```diff
@@ -0,0 +1,197 @@
+        underTest.flush();
+        for (int j = i - batch; j >= 0 && j < i; j++) {
+          Truth.assertThat(resultList.get(j).isDone()).isTrue();
+          Truth.assertThat(resultList.get(j).get()).isEqualTo(j * j);
```
Please use static imports for assertThat

igorbernstein2 commented on gax/src/test/java/com/google/api/gax/batching/v2/BatcherImplTest.java:
```diff
@@ -0,0 +1,197 @@
+
+  /** Tests accumulated element are resolved when {@link Batcher#flush()} is called. */
+  @Test
+  public void testBatchingWithCallable() throws Exception {
```
Please rename this to be more descriptive. It seems like you are testing flush here, so something like `testResultsAreResolvedAfterFlush`  would be better

igorbernstein2 commented on gax/src/test/java/com/google/api/gax/batching/v2/BatcherImplTest.java:
```diff
@@ -0,0 +1,197 @@
+    int limit = 100;
+    int batch = 10;
+    List<ApiFuture<Integer>> resultList = new ArrayList<>(limit);
+    for (int i = 0; i <= limit; i++) {
```
Why does this have to happen in a loop? it will be easier to read if you just add a single element, verify that its result is not ready, flush, verify that its ready now. And maybe do it again with a second element

igorbernstein2 commented on gax/src/test/java/com/google/api/gax/batching/v2/BatcherImplTest.java:
```diff
@@ -0,0 +1,197 @@
+      result = batcher.add(5);
+    }
+    assertThat(result.isDone()).isTrue();
+    assertThat(result.get()).isEqualTo(25);
```
It should also test that new elements cant be added

igorbernstein2 commented on gax/src/test/java/com/google/api/gax/batching/v2/BatcherImplTest.java:
```diff
@@ -0,0 +1,197 @@
+  @Test(expected = IllegalStateException.class)
+  public void testWhenNoElementAdded() throws Exception {
+    underTest.close();
+    underTest.add(1);
```
Please avoid using expected exception annotations, especially with such generic exceptions....there is nothing to prove that we got to add()

igorbernstein2 commented on gax/src/test/java/com/google/api/gax/batching/v2/BatcherImplTest.java:
```diff
@@ -0,0 +1,197 @@
+              }
+            })
+        .when(mockDescriptor)
+        .splitResponse(Mockito.<Integer>anyList(), Mockito.<SettableApiFuture<Integer>>anyList());
```
why is this mock necessary? why can't SquarerBatchingDescriptorV2 be used here?

igorbernstein2 commented on gax/src/test/java/com/google/api/gax/batching/v2/BatchingCallSettingsTest.java:
```diff
@@ -0,0 +1,126 @@
+        BatchingCallSettings.newBuilder(batchingDescriptor);
+
+    assertThat(builder.getBatchingDescriptor()).isSameAs(batchingDescriptor);
+    assertThat(builder.getBatchingSettings()).isNull();
```
I would prefer default settings

igorbernstein2 commented on gax/src/test/java/com/google/api/gax/batching/v2/BatchingCallSettingsTest.java:
```diff
@@ -0,0 +1,126 @@
+
+    assertThat(builder.getBatchingDescriptor()).isSameAs(batchingDescriptor);
+    assertThat(builder.getBatchingSettings()).isNull();
+    assertThat(builder.getRetryableCodes().size()).isEqualTo(0);
```
[isEmpty](https://truth.dev/api/latest/com/google/common/truth/IterableSubject.html#isEmpty--)

igorbernstein2 commented on gax/src/test/java/com/google/api/gax/batching/v2/BatchingCallSettingsTest.java:
```diff
@@ -0,0 +1,126 @@
+    assertThat(builder.getBatchingDescriptor()).isSameAs(batchingDescriptor);
+    assertThat(builder.getBatchingSettings()).isNull();
+    assertThat(builder.getRetryableCodes().size()).isEqualTo(0);
+    assertThat(builder.getRetrySettings()).isNotNull();
```
Maybe instantiate an instance UnaryCallableSettings and check equality?

igorbernstein2 commented on gax/src/main/java/com/google/api/gax/batching/v2/BatchingCallSettings.java:
```diff
@@ -0,0 +1,141 @@
+
+import com.google.api.core.BetaApi;
+import com.google.api.core.InternalExtensionOnly;
+import com.google.api.gax.batching.BatchingSettings;
```
Please avoid using v1 classes, or if you need to use them, please copy them into the new package

igorbernstein2 commented on gax/src/test/java/com/google/api/gax/batching/v2/BatchingCallSettingsTest.java:
```diff
@@ -0,0 +1,126 @@
+    assertThat(builder.getRetryableCodes().size()).isEqualTo(0);
+    assertThat(builder.getRetrySettings()).isNotNull();
+
+    BatchingSettings batchingSettings =
```
This PR shouldn't know anything about thresholds

igorbernstein2 commented on gax/src/test/java/com/google/api/gax/rpc/testing/FakeBatchableApi.java:
```diff
@@ -82,6 +83,10 @@ public int hashCode() {
       return result;
     }
+
+    public LabeledIntList toBuilder() {
```
a LabeledIntList is a builder for itself? Maybe you meant clone()?

rahulKQL replied:
Yes, I meant a `clone()`. Have now added a custom `clone` method.

igorbernstein2 commented on gax/src/test/java/com/google/api/gax/rpc/testing/FakeBatchableApi.java:
```diff
@@ -171,4 +176,49 @@ public long countBytes(LabeledIntList request) {
+
+    @Override
+    public long countBytes(Integer element) {
+      return String.valueOf(element).length();
```
you can probably hardcode 1

igorbernstein2 commented on gax/src/test/java/com/google/api/gax/batching/v2/BatchingCallSettingsTest.java:
```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2016 Google LLC
```
2019

rahulKQL commented on gax/src/main/java/com/google/api/gax/batching/v2/BatcherImpl.java:
```diff
@@ -0,0 +1,189 @@
+
+  /** {@inheritDoc} */
+  @Override
+  public ApiFuture<ElementResultT> add(final ElementT element) {
```
My apologies, this was an overlook.

rahulKQL commented:
@igorbernstein2 
Please have another round of review, I have addressed the feedback comments.

Also, I decided to remove `BatchingCallSettings.java` from this PR. I believe till this PR, we are not expecting any special configuration for `UnaryCallable` so a generic unaryCallable should suffice to perform batching RPCs.
Would add BatchingCallSettings along with Default Batching Settings and flow controller in followup PRs.

rahulKQL commented on gax/src/main/java/com/google/api/gax/batching/v2/BatcherImpl.java:
```diff
@@ -0,0 +1,189 @@
+        new ApiFutureCallback<ResponseT>() {
+          @Override
+          public void onSuccess(ResponseT response) {
+            batchingDescriptor.splitResponse(response, accumulatedBatch.results);
```
Updated this with a `Throwable` `catch`. Please let me know if we should add any more checks apart from this.

rahulKQL commented on gax/src/main/java/com/google/api/gax/batching/v2/BatcherImpl.java:
```diff
@@ -0,0 +1,215 @@
+ * <p>This class is not thread-safe, and expects to be used from a single thread.
+ */
+@BetaApi("The surface for batching is not stable yet and may change in the future.")
+@InternalExtensionOnly("For google-cloud-java client use only.")
```
Updated with `@InternalApi`

igorbernstein2 commented on gax/src/main/java/com/google/api/gax/batching/v2/BatcherImpl.java:
```diff
@@ -0,0 +1,219 @@
+  private void awaitAllOutstandingBatches() throws InterruptedException {
+    while (numOfOutstandingBatches.get() > 0) {
+      synchronized (flushLock) {
+        flushLock.wait(DEFAULT_WAIT_TIME_MS);
```
I'm not sure I understand why we need a timeout

rahulKQL replied:
I added this as a defense, Also inspired by existing [CBT-OperationalAccountant](https://github.com/googleapis/cloud-bigtable-client/blob/master/bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/OperationAccountant.java#L97-L102)

igorbernstein2 replied:
What is it defending against?

rahulKQL replied:
I added this for the case of RPC taking too long, which would enable this thread to keep on checking the numOfOutStandingBatch.

I now realized that this would not add any meaning. The only update we are waiting here would be from onBatchComplete's notifyAll(), which will execute on all cases.
Thanks a lot for pointing it out. I will remove this timeout.

romlogic commented on gax/src/test/java/com/google/api/gax/batching/v2/BatcherImplTest.java:
```diff
@@ -0,0 +1,214 @@
+    verify(mockUnaryCallable, times(1)).futureCall(any(LabeledIntList.class));
+  }
+
+  /** Tests results are resolves when {@link BatchingDescriptor#splitResponse} throws exception. */
```
Resolves test results when ... throws exception.

romlogic commented on gax/src/test/java/com/google/api/gax/batching/v2/BatcherImplTest.java:
```diff
@@ -0,0 +1,214 @@
+        .splitResponse(Mockito.<Integer>anyList(), Mockito.<SettableApiFuture<Integer>>anyList());
+  }
+
+  /** Tests results are resolves when {@link BatchingDescriptor#splitException} throws exception. */
```
Resolves test results when {@linkBatchingDescriptor#splitException} throws exception. 

or 

Tests results resolve when ... throws exception. (Editors note: I prefer the first sentence as it starts with an action verb.)

romlogic commented on gax/src/main/java/com/google/api/gax/batching/v2/BatcherImpl.java:
```diff
@@ -0,0 +1,219 @@
+  }
+
+  /**
+   * This class represent one logical Batch. It accumulates all the elements and it's corresponding
```
This class represents one logical batch. It accumulates all the elements and their corresponding future element results for one batch.

romlogic commented on gax/src/test/java/com/google/api/gax/rpc/testing/FakeBatchableApi.java:
```diff
@@ -171,4 +176,49 @@ public long countBytes(LabeledIntList request) {
       return Math.min(counter, 5);
```
line 174. should be "than the sum of their components."

vam-google commented on gax/src/main/java/com/google/api/gax/batching/v2/BatcherImpl.java:
```diff
@@ -0,0 +1,216 @@
+ */
+@BetaApi("The surface for batching is not stable yet and may change in the future.")
+@InternalApi
+public class BatcherImpl<ElementT, ElementResultT, RequestT, ResponseT>
```
Please specify in the doc meaning of each type argument.

rahulKQL replied:
Thanks for the feedback, Added type argument details on `BatcherImpl` as well.

vam-google commented on gax/src/main/java/com/google/api/gax/batching/v2/BatcherImpl.java:
```diff
@@ -0,0 +1,216 @@
+  private final RequestT prototype;
+  private Batch<ElementT, ElementResultT, RequestT> currentOpenBatch;
+
+  private final AtomicInteger numOfOutstandingBatches = new AtomicInteger(0);
```
The doc for this class says it is not thread-safe and is expected to be used from a single thread. Why is this using `AtomicInteger` here?

igorbernstein2 replied:
Externally its not  threadsafe and can only be used from a single thread. Internally there are 2 threads: the a single caller thread and the grpc callbacks from a different thread. Each racing to inc/dec the counter

vam-google commented on gax/src/main/java/com/google/api/gax/batching/v2/BatcherImpl.java:
```diff
@@ -0,0 +1,216 @@
+  private boolean isClosed = false;
+
+  private BatcherImpl(Builder<ElementT, ElementResultT, RequestT, ResponseT> builder) {
+    this.prototype = checkNotNull(builder.prototype, "RequestPrototype cannot be null.");
```
Please do not put period (`.`) at the end of messages like this to stay consistent with the rest of the codebase (applies to other places as well, like line 113).

Also the message probably should not use the class names to indicate what is null (either use field names i.e `prototype` instead of `RequestPrototype`, or regular langugae to describe what is null, otherwise it looks like the "class" itself is null).

vam-google commented on gax/src/main/java/com/google/api/gax/batching/v2/BatcherImpl.java:
```diff
@@ -0,0 +1,216 @@
+  }
+
+  /** Builder for a BatcherImpl. */
+  public static class Builder<ElementT, ElementResultT, RequestT, ResponseT> {
```
can this be `@Autovalue`? I'm not a big fan of `@Autovalue` but we already use it extensively in gax-java, so lets reuse it here if possible to reduce boilerplate code typing.

igorbernstein2 replied:
I don't BatcherImpl is a value object

vam-google replied:
I don't think it has to be a value object, it can be anything, which can be constructed with a builder. 
Now when I look at it, do we even need a builder here? I'm just trying to figure out if we can reduce boilerplate code here (either by generating builder with the autovalue annotation or not having it at all).

igorbernstein2 replied:
I dont think a builder is necessary here. Given the choice between AutoValue and removing the builder I would prefer removing the builder

rahulKQL replied:
I have removed the Builder from the current change and annotated this class with `@AutoValue`.

Just a side note, We may need to come back on this when introducing flow controllers. As after implementing triggers, we would be expecting 6 arguments(i.e. existing ones + BatchingSettings, FlowController, and an Executor).

igorbernstein2 replied:
TLDR version: please remove AutoValue for the time being.


Longer version:
I've given this more thought, and I really think that AutoValue is really the  wrong tool for this. In the [best practices](https://github.com/google/auto/blob/master/value/userguide/practices.md#equals-means-interchangeable) it states

> Don't use AutoValue to implement value semantics unless you really want value semantics. In particular, you should never care about the difference between two equal instances.

This is definitely not the case for BatcherImpl since a Batcher has mutable state and we do care which instance we call flush on. Also, using AutoValue here implies that equality will use value semantics and be determined by the prototype, callable & batchingDescriptor:

```java
BatcherImpl.Builder builder = ...;

Batcher b1 = builder.build();
Batcher b2 = builder.build();
b2.add(...)

b1.equals(b2) == true; 
```
Which makes no sense. I think the only meaningful equality here is referential and goes against the best practices of AutoValue.

Since you already removed the Builder until a later PR, please remove the use of AutoValue here since it buys you very little. Once that PR comes up I'll chat with vam about the best path forward.

vam-google commented on gax/src/main/java/com/google/api/gax/batching/v2/BatcherImpl.java:
```diff
@@ -0,0 +1,216 @@
+            try {
+              batchingDescriptor.splitResponse(response, accumulatedBatch.results);
+            } catch (Throwable ex) {
+              for (SettableApiFuture<ElementResultT> result : accumulatedBatch.results) {
```
Maybe move things like this to the `Batch` class itself (L202)? It already  has `add()` method for example.

rahulKQL replied:
Refactored and moved `batchingDescriptor#splitRespnse` or `batchingDescriptor#splitException` into `Batch` class.

vam-google commented on gax/src/main/java/com/google/api/gax/batching/v2/BatcherImpl.java:
```diff
@@ -0,0 +1,216 @@
+    currentOpenBatch = null;
+    numOfOutstandingBatches.incrementAndGet();
+
+    final ApiFuture<ResponseT> batchResponse =
```
If this throws an exception, `numOfOutstandingBatches` incremented above (L138) will never be decremented, is it right? Can we do something about it?

rahulKQL replied:
Thanks for pointing it out, Now incrementing the counter only after receiving `ApiFuture<ResponseT>`.

vam-google replied:
Wait, I think it can be even worse now, especially taking into account that `onBatchCompletion()` checks for 0 equality:
```java
if (numOfOutstandingBatches.decrementAndGet() == 0)
```
not
```java
if (numOfOutstandingBatches.decrementAndGet() <= 0)
```
I believe in case if that thing executes really fast, the callback can be called before the caller receives the `batchResponse` future and manages to increment the batches, so `numOfOutstandingBatches.decrementAndGet()` will return -1.

In general i had in mind some finally block or something, but the changed version looks even more dangerous than original (the possible exception was very theoretical thing, but now it is a real danger of getting -1, when 0 is expected).

igorbernstein2 replied:
The order of operations is
1. increment numOfOutstandingBatches
2. add the callback
3. at some point in the future, most likely on a different thread, invoke onBatchCompletion which calls numOfOutstandingBatches.decrementAndGet

So it doesn't matter how fast things execute, there is no race. It's possible for the callback to to be invoked before the caller gets the batchResponse future, but it doesn't matter. The count will be 0.

Please let me know what I'm missing because I dont see a problem here

rahulKQL replied:
@igorbernstein2/ @vam-google I followed Igor's last comment and I believe using a public Xor should be fine. Please have a fresh look.

vam-google commented on gax/src/main/java/com/google/api/gax/batching/v2/BatcherImpl.java:
```diff
@@ -0,0 +1,216 @@
+ * Queues up the elements until {@link #flush()} is called; once batching is over, returned future
+ * resolves.
+ *
+ * <p>This class is not thread-safe, and expects to be used from a single thread.
```
This class assigns callbacks with `directExecutor` (i.e. executor will be executed in same thread which resolved the future) and inside the callback we call this class's method `onBatchCompletion()` (in `finally` block of each callback). There is no guarantee that the callback thread and the thread instantiated this class will be same thread. It looks like this class has to be thread-safe.

igorbernstein2 replied:
See comment above. Externally we consider this class not thread safe. But internally this class handles races between the single caller thread and multiple grpc threads for callbacks

vam-google replied:
Maybe just make this class thread safe? People will be maintaining it, modifying etc. Even if everything works fine now, it will be easier to break things in the future. I do not insist, but it is my experience in general: classes are either not aware of threads whatsoever, or immutable or thread-safe. Not thread-safe (but uses threads stuff and is in general aware that it is used in concurrent), thread-compatible, thread-hostile - kind of time bombs, waiting until somebody uses them in a wrong way.

igorbernstein2 replied:
Let's take it a step at a time. Thread safety will come up again when @rahulKQL gets to implementing timed flush. There are still at least 2 PRs coming up:
1a. autoflush based on element count
1b. autoflush based on request size
2. autoflush based on time

After we get an idea as to what is required for 2, we can have a much more informed discussion about external thread safety

vam-google commented on gax/src/main/java/com/google/api/gax/batching/v2/BatcherImpl.java:
```diff
@@ -0,0 +1,216 @@
+
+  private final AtomicInteger numOfOutstandingBatches = new AtomicInteger(0);
+  private final Object flushLock = new Object();
+  private boolean isClosed = false;
```
Please make it `volatile`

igorbernstein2 replied:
This is only used by the single caller thread, so it doesn't need to be volatile

vam-google replied:
Ok, but it is practically free to make it such, ans feels safer (a little bit of defensive programming never hurts in concurrent programming).

igorbernstein2 replied:
I generally prefer to avoid defensive programming because you end up hiding the intention of the code. But I agree with you that volatile is free here. So if you feel strongly, then I won't object

vam-google replied:
I would still add volatile

igorbernstein2 replied:
@rahulKQL can you add the volatile and I will merge this PR

rahulKQL replied:
I am sorry, I haven't marked this ongoing conversation as closed.
I have already added a `volatile` for [`isClosed`](https://github.com/googleapis/gax-java/pull/716/files#diff-639d320c75b014154cb1dd32ba32e100R70) flag.

rahulKQL commented:
@vam-google, @igorbernstein2, I have tried to address your feedback, Please have a fresh look at this change.

igorbernstein2 commented:
@vam-google Thanks for the review!
